### DISCUSSION TITLE
mpy_cross for multiple versions?

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#15176] [Libraries & Drivers]
I've been using mpy_cross for v1.17 upython to squash my .py files to.mpy. The problem is lately I need to squash v1.22 .py files & I know if I download the v1.22 version I'll loose the v1.17 version.

The last time I asked this question there was talk of a universal mpy_cross that could squash any version .py file. Just wondering if that has transpired or not? If not any suggestions for how to have 2 versions of mpy_cross on the same windows PC.

**Comment by Josverl** [comment]
See https://pypi.org/project/mpy-cross-multi/

Curtesy of @dlech

**Reply by kjm1102** [reply]
thanks

**Comment by kjm1102** [comment]
The  https://pypi.org/project/mpy-cross-multi/ says it's good to run on versions >=3.10 <4.0 so I'm assuming
```
C:\Users\Admin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages>pip install mpy-cross-multi
ERROR: Could not find a version that satisfies the requirement mpy-cross-multi (from versions: none)
ERROR: No matching distribution found for mpy-cross-multi
WARNING: You are using pip version 21.3.1; however, version 24.0 is available.
```
means my Python.3.9 is 3.09?

I'm keen to not loose my v1.17 mpy-cross that runs in v3.9 python. If I upgrade to 3.10 & mpy-cross stops working &  mpy-cross-multi is a dud then I'm worse off. Has anybody actually tried mpy-cross-multi? Does it actually work?

**Reply by dlech** [reply]
mpy-cross-multi is brand new, so we are sure to have a few bugs to work out like this. It should be no problem to make it work on older Python versions.

If you have any other issues, please leave them at https://github.com/pybricks/python-mpy-cross-multi.

**Reply by dlech** [reply]
It should be working on Python >= 3.8 now (just published v1.1.2).

**Comment by Josverl** [comment]
also using pipx you could use 
 - `pipx run mpy-cross==<version>` with pipx doing the JIT downloading and installing 

```
scratch> pipx --verbose run mpy-cross==1.19.1 .\vraag.py --version
MicroPython v1.19.1 on 2022-06-18; mpy-cross emitting mpy v6
scratch> pipx --verbose run mpy-cross==1.23.0 .\vraag.py --version
MicroPython v1.23.0 on 2024-06-01; mpy-cross emitting mpy v6.3```

**Reply by kjm1102** [reply]
I normally invoke the v1.17 mpy-cross I have now with  mpy-cross -v code.py where code.py is the file to be converted. Any thoughts on the cmd I should use to do the same via pipx without re-downloading mpy-cross each time? 
```
C:\Users\Admin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\mpy_cross>pipx --verbose run mpy-cross==1.22.2
⚠️  mpy-cross is already on your PATH and installed at .\mpy-cross.EXE. Downloading and running anyway.
no input file
```

**Reply by Josverl** [reply]
In my experience just prefix the cmd with `pipx run `

Pipx does not downloadable unless it it can't find the package in it's cache


Here's a simplified explanation of how it works:
- **Check Cache**: Look for the package in the shared virtual environment.
- **Download if Needed**: If the package isn't there, download it.
- **Run from Cache**: If it's already downloaded, use the cached version.

**Reply by kjm1102** [reply]
I can't figure out how to stop downloading the .exe each time I use it 
```
C:\Users\Admin\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\mpy_cross>pipx run  mpy-cross==1.22.2 -v code.py
⚠️  mpy-cross is already on your PATH and installed at .\mpy-cross.EXE. Downloading and running anyway.
```
but hey it works and there is a lot to be said for something that actually works!

**Comment by andrewleech** [comment]
Yeah I realise this is a highly requested feature for mpy_cross. I resisted adding it to my package on pypi because it's a fully automated build and deploy and I couldn't figure out how to add back support easily without manual intervention. 

I've got it figured now though so been working on this the last few nights. I think it works pretty well now though needs some more testing before I release it: 

https://gitlab.com/alelec/mpy_cross/-/merge_requests/12

Wheels for each platform can be downloaded for testing from the pipeline archives: 
![Screenshot_20240606-183654.png](https://github.com/micropython/micropython/assets/3318786/6dc12d81-9e04-4f09-8ada-14810ad6815e)

**Comment by andrewleech** [comment]
The `mpy-cross` package I maintain on pypi has been extended to support building for old versions of micropython

https://pypi.org/project/mpy-cross/

## Older Version Support
This package also supports running mpy-cross in compatibility mode to generage mpy files for older releases of micropython. Compatibility mode can be specified either by micropython version, eg:

```
mpy-cross --compat 1.19 my_app.py
```
Or by specifying the bytecode version, eg:
```
mpy-cross --bytecode 6 my_app.py
```
## Usage

```
usage: mpy-cross [<opts>] [-X <implopt>] [--] <input filename>
Options:

-c <version> : --compat <version> : Run mpy-cross in compatibility mode for given micropython version.

-b <version> : --bytecode <version> : Output specific bytecode version for use with older micropython versions.

--version : show version information
-o : output file for compiled bytecode (defaults to input filename with .mpy extension, or stdout if input is stdin)
-s : source filename to embed in the compiled bytecode (defaults to input file)
-v : verbose (trace various operations); can be multiple
-O[N] : apply bytecode optimizations of level N

Target specific options:
-msmall-int-bits=number : set the maximum bits used to encode a small-int
-march=<arch> : set architecture for native emitter; x86, x64, armv6, armv6m, armv7m, armv7em, armv7emsp, armv7emdp, xtensa, xtensawin

Implementation specific options:
  emit={bytecode,native,viper} -- set the default code emitter
  heapsize=<n> -- set the heap size for the GC (default 2097152)
```
